packages/typegpu/src/smol/generationHelpers.ts

```diff
@@ -0,0 +1,232 @@
+  ) {
+    return undefined;
+  }
+
+  if ((isDerived(targetType) || isSlot(targetType)) && propName === 'value') {
+    const ctx = getResolutionCtx();
+    if (!ctx) {
+      throw new Error(
+        'Resolution context not found when unwrapping slot or derived',
+      );
+    }
+    return ctx.unwrap(targetType) as BaseData;
+  }
+
+  let target = targetType as AnyWgslData;
+  if ('dataType' in target) {
+    target = target.dataType as AnyWgslData;
+  }
+  while (isDecorated(target)) {
+    target = target.inner as AnyWgslData;
+  }
+  const targetTypeStr =
+    'kind' in target ? (target.kind as string) : target.type;
+
+  if (targetTypeStr === 'struct') {
+    return (target as WgslStruct).propTypes[propName];
+  }
+
+  const propLength = propName.length;
+  if (
+    swizzleableTypes.includes(
+      targetTypeStr as (typeof swizzleableTypes)[number],
+    ) &&
+    propLength >= 1 &&
+    propLength <= 4
+  ) {
+    const swizzleTypeChar = targetTypeStr[4] as SwizzleableType;
+    const swizzleType =
+      swizzleLenToType[swizzleTypeChar][propLength as SwizzleLength];
+    if (swizzleType) {
+      return swizzleType;
+    }
+  }
+
+  return isWgslData(target) ? target : undefined;
+}
+
+export function getTypeForIndexAccess(resource: Wgsl): BaseData | undefined {
+  if (typeof resource === 'string' || typeof resource === 'number') {
+    return undefined;
+  }
+
+  if (isDerived(resource) || isSlot(resource)) {
+    return getTypeForIndexAccess(resource.value as Wgsl);
+  }
+
+  if (isWgslData(resource)) {
+    // array
+    if (isWgslArray(resource)) {
+      return resource.elementType;
+    }
+
+    // vector or matrix
+    if (resource.type in indexableTypeToResult) {
+      return indexableTypeToResult[
+        resource.type as keyof typeof indexableTypeToResult
+      ];
+    }
+  }
+
+  return undefined;
+}
+
+export function getTypeFromWgsl(resource: Wgsl): AnyWgslData | UnknownData {
+  if (isDerived(resource) || isSlot(resource)) {
+    return getTypeFromWgsl(resource.value as Wgsl);
+  }
+
+  if (typeof resource === 'string') {
+    return UnknownData;
+  }
+  if (typeof resource === 'number') {
+    return numericLiteralToResource(String(resource))?.dataType ?? UnknownData;
+  }
+  if (typeof resource === 'boolean') {
+    return bool;
+  }
+
+  return isWgslData(resource) ? resource : UnknownData;
+}
+
+export function numericLiteralToResource(value: string): Resource | undefined {
+  // Hex literals
+  if (/^0x[0-9a-f]+$/i.test(value)) {
+    return { value, dataType: abstractInt };
+  }
+
+  // Binary literals
+  if (/^0b[01]+$/i.test(value)) {
+    return {
+      value: `${Number.parseInt(value.slice(2), 2)}`,
+      dataType: abstractInt,
+    };
+  }
+
+  // Floating point literals
+  if (/^-?(?:\d+\.\d*|\d*\.\d+)$/i.test(value)) {
+    return { value, dataType: abstractFloat };
+  }
+
+  // Floating point literals with scientific notation
+  if (/^-?\d+\.\d+e-?\d+$/i.test(value)) {
```

Comment:
I think your force push removed my changes.
```suggestion
  if (/^-?\d+(?:\.\d+)?e-?\d+$/i.test(value)) {
```